### DISCUSSION TITLE
confused about the websocket/chat example model

### DISCUSSION THREAD

**Discussion by yatesco** [tokio-rs/axum#604] [Q&A] [answered]
<websocket newbie>
Does the web socket loop persist as long as the client is connected, which I can't believe it does do to resource drainage. If not, how do I associate a "user" (e.g. their session id/cookie) with their particular socket?

Do I split the socket on_connect and then shove the write site into a map? I guess it needs to be a weak reference in case the client disappears?

I wasn't kidding about the newbie section ;-).

**Comment by davidpdrsn** [comment, marked as answer]
Does the chat example help https://github.com/tokio-rs/axum/blob/main/examples/chat/src/main.rs?

**Reply by yatesco** [reply]
Not really - I've read it about 3 times :-). I'd claim the problem is me, though, not the example, but I think I understand that example.

Imagine I want to be able to message individual users. I'd imagine I would have a map from userId (e.g. username) to the `sender` from `let (mut sender, mut receiver) = stream.split();` at line 61. 

I guess my question is, does the `websocket` fn run for as long as the client is connected? If so, then fine, I can wrap up a "userRegistry" in an Arc and pass it in.

I think I'm just overthinking it.

**Reply by davidpdrsn** [reply]
The function runs until it returns 😅 it's passed directly to `tokio::spawn` so its really up to you. Could be when the client disconnects (ie receiving a message fails) but you're free to do other stuff afterwards, or return early if you want.

How exactly you want to manage the connections and whatnot I don't have recommendations for. A HashMap and some tokio mpsc channels is probably a good way to start.

**Reply by yatesco** [reply]
right :-), so just to be clear - the loop at line 108:
```rust
    // This task will receive messages from client and send them to broadcast subscribers.
    let mut recv_task = tokio::spawn(async move {
        while let Some(Ok(Message::Text(text))) = receiver.next().await {
            // Add username before message.
            let _ = tx.send(format!("{}: {}", name, text));
        }
    });
``` 
will run for a long time until the client disconnects, if I'm reading this right?

**Reply by davidpdrsn** [reply]
Yep that's correct!

**Reply by yatesco** [reply]
thanks - I knew I was overthinking it ;-)

**Reply by davidpdrsn** [reply]
No worries. Happy to help 😊
